design/volume-group-snapshot.md

```diff
@@ -0,0 +1,428 @@
+    - Users will have to label the PVCs before invoking the backup operation.
+    - This label would act as a group identifier for the PVCs to be grouped under a specific VGS.
+    - It will be used to collect the PVCs to be used for a particular instance of VGS object.
+
+- Changes to the Existing PVC ItemBlockAction plugin:
+    - Currently the PVC IBA plugin is applied to PVCs and adds the RelatedItems for the particular PVC into the ItemBlock.
+    - At first it checks whether the PVC is bound and VolumeName is non-empty.
+    - Then it adds the related PV under the list of relatedItems.
+    - Following on, the plugin adds the pods mounting the PVC as relatedItems.
+    - Now we need to extend this PVC IBA plugin to add the PVCs to be grouped for a particular VGS object, so that they are processed together under an ItemBlock by Velero.
+        - First we will check if the PVC that is being processed by the plugin has the user specified VGS label.
+        - If it is present then we will execute a List call in the namespace with the label as a matching criteria and see if this results in any PVCs (other than the current one).
+        - If there are PVCs matching the criteria then we add the PVCs to the relatedItems list.
+        - This helps in building the ItemBlock we need for VGS processing, i.e. we have the relevant pods and PVCs in the ItemBlock.
+
+- Updates to CSI PVC plugin:
+    - This is the plugin which creates VolumeSnapshots for the CSI PVCs if required.
+    - We will have to update this plugin to:
+        - Create VGS for the group of PVCs
+            - At First we will check if PVC has the user specified VGS label.
+            - If yes, then it implies it needs to be part of a VGS instance.
+            - Check if VGS exists for this PVC or not, we don't want to recreate VGS object.
+            - We can check if VGS exists by listing the VGS in the namespace using the backup UUID alongwith the label used for the VGS (we will be creating VGS with backup UUID and user specified VGS  group label) and see if the PVC is part of any of the existing VGS for the particular backup operation we are currently in.
+            - If not then create the VGS object with the required labels (in this case backup UUID and the user specified VGS group label) and also add the PVC names (that are included in the VGS) as annotations. (This will be useful for our VGS restore operation where we would need PVC resource identifiers).
+            - Once VGS is created, add the VGS as an additional item to be included in the backup.
+        - Skip creation of VS for PVC that are already included in VGS
+            - Now if the PVC has the user specified VGS label that means it is destined to be part of some VGS and the VS would be created by the K8s VGS controller and Velero does not need to explicitly create the VS for the PVC.
+
+- Add a new VolumeGroupSnapshot(VGS) BIA plugin
```

Comment:
Please add more details to elaborate on how things diff between CSI snapshot backup and CSI snapshot data movement, below is some I can come out for now:
1. For data mover backup, waiting for VS to get to `readyToUse` is done by the data mover exposer. The CSI plugin only waits for the snapshotHandle.
2. For data mover backup, dataUpload CRs are created in the PVC BIA, after creating the snapshot and snapshot handle appears. The VS plugin does nothing.
3. For data mover backup, VGSClass is not required to be backed up, as same as other snapshot objects, i.e., VS and VSC

I will think more about this later, but right now, I think for data mover backup, we should bypass the VGS BIA, just like we bypass VS and VSC BIAs for data mover as the current behavior.

Reply:
> Is there an equivalent for the VSC of the snapHandle being present

At present, data mover bypasses BIA of VSC, so doing things in BIA of VSC doesn't make any differences.